src/impy/common.py

```diff
@@ -252,28 +294,51 @@ def to_hepmc3(self, genevent=None):
         """
         import pyhepmc  # delay import
 
+        model, version = self.generator
+
         if genevent is None:
             genevent = pyhepmc.GenEvent()
+            genevent.run_info = pyhepmc.GenRunInfo()
+            genevent.run_info.tools = [(model, version, "")]
+
+        # We must apply some workarounds so that HepMC3 conversion and IO works
+        # for all models. This should be revisited once the fundamental issues
+        # with particle histories have been fixed.
+        if model == "Pythia" and version.startswith("8"):
+            # must deselect parton showers in Pythia-8 to use HepMC3 IO
+            mask = True
+            apid = np.abs(self.pid)
+            for pid in quarks_and_diquarks_and_gluons:
+                mask &= apid != pid
+            ev = self[mask]
+        elif model in ("UrQMD", "PhoJet"):
+            # can only save final state until history is fixed
+            warnings.warn(
+                f"{model}-{version}: only final state particles "
+                "available in HepMC3 event",
+                RuntimeWarning,
+            )
+            ev = self.final_state()
+        else:
+            ev = self
```

Comment:
This is not elegant, because we apply the workaround in common.py, while all model specific stuff should be in the respective submodule, e.g. pythia8.py. I chose this solution because I hope it is temporary, but I am open to make this proper by adding a hook here that calls into a private method that a Model class can customize to implement these workarounds.

Reply:
Actually, I overlooked this. In Phojet particle history should be fine. It fails in DPMJET from the discussion here #74 .